server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.james.mailrepository.api.MailKey;
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.mailrepository.api.MailRepositoryStore;
+import org.apache.james.mailrepository.api.MailRepositoryUrl;
+import org.apache.james.util.streams.Iterators;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class MailRepositoryBasedDeletedMessageVault implements DeletedMessageVault {
+    public static class Configuration {
+        private final String urlPrefix;
+
+        public Configuration(String urlPrefix) {
+            this.urlPrefix = urlPrefix;
+        }
+    }
+
+    private final MailRepositoryStore mailRepositoryStore;
+    private final Configuration configuration;
+    private final MailAdapter mailAdapter;
+
+    public MailRepositoryBasedDeletedMessageVault(MailRepositoryStore mailRepositoryStore, Configuration configuration, MailAdapter mailAdapter) {
+        this.mailRepositoryStore = mailRepositoryStore;
+        this.configuration = configuration;
+        this.mailAdapter = mailAdapter;
+    }
+
+    Mono<MailRepository> repositoryForUser(User user) {
+        MailRepositoryUrl mailRepositoryUrl = MailRepositoryUrl.from(configuration.urlPrefix + user.asString());
+        return Mono.fromCallable(() -> mailRepositoryStore.select(mailRepositoryUrl));
+    }
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+
+        return repositoryForUser(user)
+            .flatMap(mailRepository -> Mono.fromCallable(() -> mailRepository.store(mailAdapter.toMail(deletedMessage))))
+            .then()
+            .publishOn(Schedulers.elastic());
+    }
+
+    @Override
+    public Publisher<Void> delete(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        MailKey mailKey = new MailKey(messageId.serialize());
+
+        return repositoryForUser(user)
+            .publishOn(Schedulers.elastic())
+            .flatMap(repository -> Mono.fromRunnable(
+                Throwing.runnable(() -> repository.remove(mailKey))));
+    }
+
+    @Override
+    public Publisher<DeletedMessage> search(User user, Query query) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(query);
+        Preconditions.checkArgument(query.getCriteria().isEmpty(), "Search is not supported yet...");
+
+        Mono<MailRepository> mailRepositoryMono = repositoryForUser(user).cache();
+
+        return mailRepositoryMono
+            .flatMapMany(Throwing.function(repository -> Flux.fromStream(Iterators.toStream(repository.list()))))
```

Comment:
can't you use `Flux.fromIterable` ?

Reply:
Not from an iterator, very sadly...

Maybe we would gain having a `Iterators.toFlux`  method?

Reply:
ok